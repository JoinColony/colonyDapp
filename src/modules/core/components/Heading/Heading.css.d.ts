export const main: string;
export const themePrimary: string;
export const themeDark: string;
export const themeGrey: string;
export const themeInvert: string;
export const themeUppercase: string;
export const marginNone: string;
export const marginSmall: string;
export const marginDouble: string;
export const sizeTiny: string;
export const sizeSmall: string;
export const sizeNormal: string;
export const sizeMedium: string;
export const sizeLarge: string;
export const sizeHuge: string;
export const weightThin: string;
export const weightMedium: string;
export const weightBold: string;
