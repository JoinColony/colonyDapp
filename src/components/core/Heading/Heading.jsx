/* @flow */

import type { Node } from 'react';
import type { IntlShape, MessageDescriptor, MessageValues } from 'react-intl';

import React from 'react';
import { injectIntl } from 'react-intl';

import { getMainClasses } from '~utils/css';

import styles from './Heading.css';

const displayName = 'Heading';

type Appearance = {
  theme?: 'primary' | 'dark' | 'invert' | 'uppercase',
  margin?: 'none' | 'small' | 'double',
  size: 'tiny' | 'small' | 'normal' | 'medium' | 'large' | 'huge',
  weight?: 'thin' | 'medium' | 'bold',
};

type Props = {
  /** Appearance object */
  appearance?: Appearance,
  /** String that will hard set the heading element to render */
  tagName?: string,
  /** Used to extend the functionality of the component. This will not generate a title attribute on the element. */
  children?: Node,
  /** A string or a `MessageDescriptor` that make up the headings's text */
  text?: MessageDescriptor | string,
  /** Values for text (react-intl interpolation) */
  textValues?: MessageValues,
  /** @ignore injected by `react-intl` */
  intl: IntlShape,
};

const getText = (
  children?: Node,
  text?: MessageDescriptor | string,
  textValues,
  { formatMessage }: IntlShape,
) => {
  if (children) {
    return children;
  }
  if (!text) {
    return '';
  }
  if (typeof text == 'string') {
    return text;
  }
  return formatMessage(text, textValues);
};

const Heading = ({
  appearance = { size: 'huge' },
  children,
  intl,
  tagName,
  text,
  textValues,
  ...props
}: Props) => {
  const { size } = appearance;
  const HeadingElement =
    tagName ||
    {
      huge: 'h1',
      large: 'h2',
      medium: 'h3',
      normal: 'h4',
      small: 'h5',
      tiny: 'h6',
    }[size || 'huge'];
  const value = getText(children, text, textValues, intl);
  return (
    <HeadingElement
      title={value}
      className={getMainClasses(appearance, styles)}
      {...props}
    >
      {value || children}
    </HeadingElement>
  );
};

Heading.displayName = displayName;

export default injectIntl(Heading);
