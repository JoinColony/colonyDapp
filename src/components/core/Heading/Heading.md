## Headings

### Normal headings

```js
<Heading appearance={{ size: 'huge' }}>Huge heading</Heading>
<Heading appearance={{ size: 'large' }}>Large heading</Heading>
<Heading appearance={{ size: 'medium' }}>Medium heading</Heading>
<Heading appearance={{ size: 'normal' }}>Normal heading</Heading>
<Heading appearance={{ size: 'small' }}>Small heading</Heading>
<Heading appearance={{ size: 'tiny' }}>Tiny heading</Heading>
```

### Invert headings

```js
<div style={{ backgroundColor: 'black' }}>
  <Heading appearance={{ theme: 'invert', size: 'huge' }}>Huge heading</Heading>
  <Heading appearance={{ theme: 'invert', size: 'large' }}>Large heading</Heading>
  <Heading appearance={{ theme: 'invert', size: 'medium' }}>Medium heading</Heading>
  <Heading appearance={{ theme: 'invert', size: 'normal' }}>Normal heading</Heading>
  <Heading appearance={{ theme: 'invert', size: 'small' }}>Small heading</Heading>
  <Heading appearance={{ theme: 'invert', size: 'tiny' }}>Tiny heading</Heading>
</div>
```

### Other heading themes

```js
<Heading appearance={{ theme: 'primary', size: 'huge' }}>Primary heading</Heading>
<Heading appearance={{ theme: 'dark', size: 'huge' }}>Dark heading</Heading>
<Heading appearance={{ theme: 'uppercase', size: 'huge' }}>Uppercase heading</Heading>
```

### Weights

```js
<Heading appearance={{ theme: 'dark', weight: 'bold', size: 'large' }}>Bold heading</Heading>
<Heading appearance={{ theme: 'dark', weight: 'medium', size: 'large' }}>Medium heading</Heading>
<Heading appearance={{ theme: 'dark', weight: 'thin', size: 'large' }}>Thin heading</Heading>
```
