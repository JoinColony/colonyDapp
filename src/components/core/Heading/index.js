/* @flow */

export { default } from './Heading.jsx';
