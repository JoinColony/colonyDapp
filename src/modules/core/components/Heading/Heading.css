.main {
  margin-bottom: 20px;
  font-family: var(--family-primary);
  font-size: var(--size-medium-l);
  font-weight: var(--weight-bold);
  line-height: 1.5;
  word-break: normal;
  white-space: pre-line;
  color: var(--text);
}

/*
 * Theme-ing
 *
 * Next set of classes will define our various heading themes.
 */

/*
 * Primary is using the `colony` `green` color, as opposes as a `dark-grey` color for the heading.
 * This is done for consistency: everywhere we refer to something as `primary` we expect `colony` `green`.
 */
.themePrimary {
  composes: main;
  color: var(--primary);
}

.themeDark {
  composes: main;
  color: var(--dark);
}

.themeInvert {
  composes: main;
  color: var(--text-invert);
}

.themeUppercase {
  composes: main;
  text-transform: uppercase;
  color: var(--temp-grey-11);
}

/*
 * Margins
 *
 * Next set of classes will define the headings's margins
 */

.marginNone {
  /* Need to overwrite the size class */
  margin-bottom: 0 !important;
}

.marginSmall {
  margin-bottom: 6px !important;
}

.marginDouble {
  margin-bottom: 40px;
}

/*
 * Sizing
 *
 * Next set of classes will define the headings's size
 */

.sizeTiny {
  font-size: var(--size-tiny);
}

.sizeSmall {
  font-size: var(--size-small);
}

.sizeNormal {
  font-size: var(--size-normal);
}

.sizeMedium {
  font-size: var(--size-medium-l);
  letter-spacing: 1px;
}

.sizeLarge {
  font-size: var(--size-large);
}

.sizeHuge {
  font-size: var(--size-huge);
}

/*
 * Font weight
 *
 * Next set of classes will define the headings's weight
 */
.weightThin {
  font-weight: var(--weight-normal);
}

.weightMedium {
  font-weight: var(--weight-bold);
}

.weightBold {
  font-weight: var(--weight-bold);
}
