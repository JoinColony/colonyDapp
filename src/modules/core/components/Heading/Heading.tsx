import React, { HTMLAttributes, ReactNode, useMemo } from 'react';
import { MessageDescriptor, useIntl } from 'react-intl';

import { SimpleMessageValues } from '~types/index';
import { getMainClasses } from '~utils/css';

import styles from './Heading.css';

const displayName = 'Heading';

export type Appearance = {
  theme?: 'primary' | 'dark' | 'invert' | 'uppercase' | 'grey';
  margin?: 'none' | 'small' | 'double';
  size: 'tiny' | 'small' | 'normal' | 'medium' | 'large' | 'huge';
  weight?: 'thin' | 'medium' | 'bold';
};

interface Props extends HTMLAttributes<HTMLHeadingElement> {
  /** Appearance object */
  appearance?: Appearance;

  /** String that will hard set the heading element to render */
  tagName?: string;

  /** Used to extend the functionality of the component. This will not generate a title attribute on the element. */
  children?: ReactNode;

  /** A string or a `MessageDescriptor` that make up the headings's text */
  text?: MessageDescriptor | string;

  /** Values for text (react-intl interpolation) */
  textValues?: SimpleMessageValues;
}

const Heading = ({
  appearance = { size: 'huge' },
  children,
  tagName,
  text,
  textValues,
  ...props
}: Props) => {
  const { formatMessage } = useIntl();
  const { size } = appearance;
  const HeadingElement: any =
    tagName ||
    {
      huge: 'h1',
      large: 'h2',
      medium: 'h3',
      normal: 'h4',
      small: 'h5',
      tiny: 'h6',
    }[size || 'huge'];
  const value = useMemo(() => {
    if (children) {
      return children;
    }
    if (!text) {
      return '';
    }
    if (typeof text === 'string') {
      return text;
    }
    return formatMessage(text, textValues);
  }, [children, formatMessage, text, textValues]);
  return (
    <HeadingElement // If `value` is of type `Node` (i.e. children prop), don't add broken title.
      title={typeof value === 'string' ? value : null}
      className={getMainClasses(appearance, styles)}
      {...props}
    >
      {value}
    </HeadingElement>
  );
};

Heading.displayName = displayName;

export default Heading;
