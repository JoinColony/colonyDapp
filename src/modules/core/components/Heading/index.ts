export { default, Appearance } from './Heading';
