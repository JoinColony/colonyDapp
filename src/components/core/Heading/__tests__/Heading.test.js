/* eslint-env jest */

import React from 'react';
import toJson from 'enzyme-to-json';

import { shallowWithIntl, mountWithIntl } from 'testutils';

import Heading from '../Heading.jsx';

describe('Heading component', () => {
  const mockTitle = 'Hello Rick';
  test('Renders initial component', () => {
    const wrapper = shallowWithIntl(<Heading text={mockTitle} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
  test('Works with MessageDescriptors', () => {
    const wrapper = shallowWithIntl(
      <Heading text={{ id: 'rick', defaultMessage: mockTitle }} />,
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });
  test('Shows the title value', () => {
    const wrapper = shallowWithIntl(<Heading text={mockTitle} />);
    expect(wrapper.html()).toContain(mockTitle);
  });
  test('Has default heading element props', () => {
    const wrapper = mountWithIntl(<Heading text={mockTitle} />);
    expect(wrapper.html()).toContain('<h1');
  });
  test('Correctly generates heading elements', () => {
    /*
     * Huge
     */
    let wrapper = mountWithIntl(
      <Heading appearance={{ size: 'huge' }} text={mockTitle} />,
    );
    expect(wrapper.html()).toContain('<h1');
    /*
     * Large
     */
    wrapper = mountWithIntl(
      <Heading appearance={{ size: 'large' }} text={mockTitle} />,
    );
    expect(wrapper.html()).toContain('<h2');
    /*
     * Medium
     */
    wrapper = mountWithIntl(
      <Heading appearance={{ size: 'medium' }} text={mockTitle} />,
    );
    expect(wrapper.html()).toContain('<h3');
    /*
     * Normal
     */
    wrapper = mountWithIntl(
      <Heading appearance={{ size: 'normal' }} text={mockTitle} />,
    );
    expect(wrapper.html()).toContain('<h4');
    /*
     * Small
     */
    wrapper = mountWithIntl(
      <Heading appearance={{ size: 'small' }} text={mockTitle} />,
    );
    expect(wrapper.html()).toContain('<h5');
    /*
     * Tiny
     */
    wrapper = mountWithIntl(
      <Heading appearance={{ size: 'tiny' }} text={mockTitle} />,
    );
    expect(wrapper.html()).toContain('<h6');
  });
  test('Can overwrite the heading element', () => {
    const eliteElement = 'h1337';
    const wrapper = shallowWithIntl(
      <Heading tagName={eliteElement} text={mockTitle} />,
    );
    expect(wrapper.html()).toContain(`<${eliteElement}`);
  });
});
