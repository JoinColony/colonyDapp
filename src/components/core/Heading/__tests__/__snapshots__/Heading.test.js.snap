// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Heading component Renders initial component 1`] = `
<Heading
  intl={
    Object {
      "defaultFormats": Object {},
      "defaultLocale": "en",
      "formatDate": [Function],
      "formatHTMLMessage": [Function],
      "formatMessage": [Function],
      "formatNumber": [Function],
      "formatPlural": [Function],
      "formatRelative": [Function],
      "formatTime": [Function],
      "formats": Object {},
      "formatters": Object {
        "getDateTimeFormat": [Function],
        "getMessageFormat": [Function],
        "getNumberFormat": [Function],
        "getPluralFormat": [Function],
        "getRelativeFormat": [Function],
      },
      "locale": "en",
      "messages": Object {},
      "now": [Function],
      "onError": [Function],
      "textComponent": "span",
      "timeZone": null,
    }
  }
  text="Hello Rick"
/>
`;

exports[`Heading component Works with MessageDescriptors 1`] = `
<Heading
  intl={
    Object {
      "defaultFormats": Object {},
      "defaultLocale": "en",
      "formatDate": [Function],
      "formatHTMLMessage": [Function],
      "formatMessage": [Function],
      "formatNumber": [Function],
      "formatPlural": [Function],
      "formatRelative": [Function],
      "formatTime": [Function],
      "formats": Object {},
      "formatters": Object {
        "getDateTimeFormat": [Function],
        "getMessageFormat": [Function],
        "getNumberFormat": [Function],
        "getPluralFormat": [Function],
        "getRelativeFormat": [Function],
      },
      "locale": "en",
      "messages": Object {},
      "now": [Function],
      "onError": [Function],
      "textComponent": "span",
      "timeZone": null,
    }
  }
  text={
    Object {
      "defaultMessage": "Hello Rick",
      "id": "rick",
    }
  }
/>
`;
